[JSC] Make StructureTransitionTable link half-weak

Constellation opened the pull request:
#### e5a9344fbb6d6e72bf8250e8215b50da8528ffe5
<pre>
[JSC] Make StructureTransitionTable link half-weak
<a href="https://bugs.webkit.org/show_bug.cgi?id=255137">https://bugs.webkit.org/show_bug.cgi?id=255137</a>
rdar://107739562

Reviewed by Mark Lam.

This patch optimizes Object.assign by doing two things.

1. Add putOwnDataPropertyBatching. We batch the property additions by just tracing structure transitions.
   This offers a chance to remove unnecessary intermediate butterfly allocations.
2. We found that now the hottest code in Object.assign is Structure transition itself. This patch optimizes
   this by rearchitect structure transition table mechanism. Previously we were having Weak&lt;Structure&gt; for
   single transition. But this is (1) too costly, Weak is generic and not so efficient abstraction. Every
   access to this field becomes double-pointer loading. And (2) too aggressive. Most of Structure lives for
   a long time while instances die quickly. We should allow keeping this structure chain alive at least during
   Eden collections. To make both work, this patch changes this link &quot;half-weak&quot;. During Eden collection, we
   always mark it. But in full collection, we ignore this link, and GC End phase clears it if this Structure
   is collected. This way allows us to have good balance of Structure book-keeping, plain simple field access
   for structure transition, and not doing GC End phase clearing for Structures in each Eden GC.

Object.assign gets 30% better.
                                         ToT                     Patched

    object-assign-multiple        134.7853+-0.4586     ^    103.1172+-0.1968        ^ definitely 1.3071x faster

* JSTests/microbenchmarks/object-assign-multiple.js: Added.
(extend1):
(test):
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::finalizeUnconditionally):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeUnconditionally):
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::UnlinkedFunctionExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalizeMarkedUnconditionalFinalizers):
(JSC::Heap::finalizeUnconditionalFinalizers):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/ArgList.h:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/FunctionExecutableInlines.h:
(JSC::FunctionExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/GlobalExecutable.cpp:
(JSC::GlobalExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/GlobalExecutable.h:
* Source/JavaScriptCore/runtime/InferredValue.h:
* Source/JavaScriptCore/runtime/InferredValueInlines.h:
(JSC::InferredValue&lt;JSCellType&gt;::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSWeakObjectRef.cpp:
(JSC::JSWeakObjectRef::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.h:
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::setSingleTransition):
(JSC::StructureTransitionTable::contains const):
(JSC::StructureTransitionTable::add):
(JSC::Structure::addNewPropertyTransition):
(JSC::Structure::removeNewPropertyTransition):
(JSC::Structure::attributeChangeTransition):
(JSC::Structure::nonPropertyTransitionSlow):
(JSC::Structure::visitChildrenImpl):
(JSC::Structure::setBrandTransition):
(JSC::Structure::finalizeUnconditionally):
(): Deleted.
(JSC::singleSlotTransitionWeakOwner): Deleted.
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::shouldDoCacheableDictionaryTransitionForAdd):
(JSC::Structure::transitionCountEstimate const):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::addPropertyTransitionToExistingStructureImpl):
(JSC::StructureTransitionTable::trySingleTransition const):
(JSC::StructureTransitionTable::get const):
(JSC::StructureTransitionTable::finalizeUnconditionally):
(JSC::StructureTransitionTable::singleTransition const): Deleted.
* Source/JavaScriptCore/runtime/StructureRareData.cpp:
(JSC::StructureRareData::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/StructureRareData.h:
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::StructureTransitionTable::~StructureTransitionTable):
(JSC::StructureTransitionTable::setMap):
(JSC::StructureTransitionTable::StructureTransitionTable): Deleted.
(JSC::StructureTransitionTable::weakImpl const): Deleted.
* Source/JavaScriptCore/runtime/SymbolTable.h:
* Source/JavaScriptCore/runtime/SymbolTableInlines.h:
(JSC::SymbolTable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::finalizeUnconditionally):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.cpp:
(JSC::JSWebAssemblyModule::finalizeUnconditionally):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.h:

Canonical link: <a href="https://commits.webkit.org/262809@main">https://commits.webkit.org/262809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4bda6e735ac366c60c78079a5c52492d17f6e393

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2667 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2669 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3045 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2764 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2349 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2691 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3141 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2385 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3837 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/634 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2370 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2231 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2223 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2750 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2386 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3592 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2549 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2421 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2204 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2744 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2424 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2373 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/642 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2410 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2803 "Built successfully") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/313 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2564 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/757 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/StructureInlines.h:
```diff
@@ -733,25 +740,32 @@ ALWAYS_INLINE Structure* Structure::addPropertyTransitionToExistingStructureConc
-            return jsCast<Structure*>(impl->jsValue().asCell());
-    }
+    uintptr_t pointer = m_data;
+    if (pointer & UsingSingleSlotFlag)
```
Let's use `isUsingSingleSlot()` here to be consistent.

Constellation replied:
This is intentional. This function can be called concurrently for GC marking. Thus, we need to avoid the  case, that m_data gets rewritten after using `isUsingSingleSlot()`.

MenloDorian replied:
Ah, right.  I missed that this is caching `m_data` in `pointer`.

MenloDorian commented on Source/JavaScriptCore/runtime/StructureInlines.h:
```diff
@@ -733,25 +740,32 @@ ALWAYS_INLINE Structure* Structure::addPropertyTransitionToExistingStructureConc
 
+inline void StructureTransitionTable::finalizeUnconditionally(VM& vm, CollectionScope)
+{
+    if (isUsingSingleSlot()) {
+        auto* transition = trySingleTransition();
+        if (transition && !vm.heap.isMarked(transition))
+            m_data = UsingSingleSlotFlag;
+    }
```
You can just do this instead:
```
    if (auto* transition = trySingleTransition()) {
        if (!vm.heap.isMarked(transition))
            m_data = UsingSingleSlotFlag;
    }
```

Constellation replied:
Changed.

MenloDorian commented on Source/JavaScriptCore/runtime/StructureInlines.h:
```diff
@@ -712,10 +713,16 @@ inline Structure* Structure::addPropertyTransitionToExistingStructureImpl(Struct
-    if (Structure* existingTransition = structure->m_transitionTable.get(uid, attributes, TransitionKind::PropertyAddition)) {
-        validateOffset(existingTransition->transitionOffset(), existingTransition->inlineCapacity());
-        offset = existingTransition->transitionOffset();
-        return existingTransition;
+    Structure* existingTransition = nullptr;
+    {
+        {
+            existingTransition = structure->m_transitionTable.get(uid, attributes, TransitionKind::PropertyAddition);
+        }
+        if (existingTransition) {
+            validateOffset(existingTransition->transitionOffset(), existingTransition->inlineCapacity());
+            offset = existingTransition->transitionOffset();
+            return existingTransition;
+        }
```
Why is this change needed at all?  It appears to be doing exactly the same thing.  Let's undo this.

Constellation replied:
Removed.

MenloDorian commented on Source/JavaScriptCore/runtime/StructureInlines.h:
```diff
@@ -35,6 +35,7 @@
 #include "Structure.h"
 #include "StructureChain.h"
 #include "StructureRareDataInlines.h"
+#include "SuperSampler.h"
```
This is not needed, right?  Can we remove it?

Constellation replied:
Removed.

MenloDorian commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -324,10 +324,49 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
+                JSObject* source = asObject(sourceValue);
+                source->structure()->forEachProperty(vm, [&](const PropertyTableEntry& entry) -> bool {
+                    if (entry.attributes() & PropertyAttribute::DontEnum)
+                        return true;
```
nit: We should really change this to return `IterationStatus`, but we can do that later.  Let's see, true means `IterationStatus::Continue` here.

MenloDorian commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -324,10 +324,49 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
     bool targetCanPerformFastPut = jsDynamicCast<JSFinalObject*>(target) && target->canPerformFastPutInlineExcludingProto() && target->isStructureExtensible();
+    unsigned argsCount = callFrame->argumentCount();
+
+    if (argsCount > 2 && argsCount < 5 && targetCanPerformFastPut) {
```
Can you add a comment about why the selection of argCount between 2 and 5?  This is just a heuristic, right?  As in 1 means nothing to assign, at least 2 args before there is something to assign, and less than 5 limit the number of objects we'll need to scan this way and possibly fail out to the slow path anyway, correct?

Constellation replied:
Right. 2 is `Object.assign(target, source)`. We do not need batching of multiple objects. If we increase this too much, we need to keep huge buffer, and we cut this with 5 threshold.

MenloDorian commented on Source/JavaScriptCore/runtime/JSObject.cpp:
```diff
@@ -40,6 +40,7 @@
 #include "PropertyDescriptor.h"
 #include "PropertyNameArray.h"
 #include "ProxyObject.h"
+#include "SuperSampler.h"
```
This is not needed, right?  Can  we remove it?

Constellation replied:
Removed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSObject.cpp:
```diff
@@ -4078,4 +4079,74 @@ TransitionKind JSObject::suggestedArrayStorageTransition() const
+{
+    unsigned i = 0;
+    Structure* structure = this->structure();
+    if (!(structure->isDictionary() || (structure->transitionCountEstimate() + size) > Structure::s_maxTransitionLength || !structure->canPerformFastPropertyEnumeration() || (structure->transitionWatchpointSet().isBeingWatched() && structure->transitionWatchpointSet().isStillValid()))) {
```
Are we only comparing `structure->transitionCountEstimate() + size)` against `Structure::s_maxTransitionLength` because it is smaller (and therefore, more conservative than) `s_maxTransitionLengthForNonEvalPutById`?  If so, I recommend that you add a `static_assert(Structure::s_maxTransitionLength < Structure::s_maxTransitionLengthForNonEvalPutById, "We should use the more conservative max transition length")` above here to document this.

Constellation replied:
Sounds good, changed.

Constellation replied:
Ah, we do not need s_maxTransitionLengthForNonEvalPutById. We always use s_maxTransitionLength since this is for C++ usage (not from JS function). So we can just use `s_maxTransitionLength`

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Make StructureTransitionTable link half-weak
https://bugs.webkit.org/show_bug.cgi?id=255137
rdar://107739562

Reviewed by Mark Lam.

This patch optimizes Object.assign by doing two things.

1. Add putOwnDataPropertyBatching. We batch the property additions by just tracing structure transitions.
   This offers a chance to remove unnecessary intermediate butterfly allocations.
2. We found that now the hottest code in Object.assign is Structure transition itself. This patch optimizes
   this by rearchitect structure transition table mechanism. Previously we were having Weak<Structure> for
   single transition. But this is (1) too costly, Weak is generic and not so efficient abstraction. Every
   access to this field becomes double-pointer loading. And (2) too aggressive. Most of Structure lives for
   a long time while instances die quickly. We should allow keeping this structure chain alive at least during
   Eden collections. To make both work, this patch changes this link "half-weak". During Eden collection, we
   always mark it. But in full collection, we ignore this link, and GC End phase clears it if this Structure
   is collected. This way allows us to have good balance of Structure book-keeping, plain simple field access
   for structure transition, and not doing GC End phase clearing for Structures in each Eden GC.

Object.assign gets 30% better.
                                         ToT                     Patched

    object-assign-multiple        134.7853+-0.4586     ^    103.1172+-0.1968        ^ definitely 1.3071x faster

* JSTests/microbenchmarks/object-assign-multiple.js: Added.
(extend1):
(test):
* Source/JavaScriptCore/builtins/BuiltinExecutables.cpp:
(JSC::BuiltinExecutables::finalizeUnconditionally):
* Source/JavaScriptCore/builtins/BuiltinExecutables.h:
* Source/JavaScriptCore/bytecode/CodeBlock.cpp:
(JSC::CodeBlock::finalizeUnconditionally):
* Source/JavaScriptCore/bytecode/CodeBlock.h:
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.cpp:
(JSC::UnlinkedFunctionExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/bytecode/UnlinkedFunctionExecutable.h:
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalizeMarkedUnconditionalFinalizers):
(JSC::Heap::finalizeUnconditionalFinalizers):
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/runtime/ArgList.h:
* Source/JavaScriptCore/runtime/ErrorInstance.cpp:
(JSC::ErrorInstance::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/ErrorInstance.h:
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/FunctionExecutableInlines.h:
(JSC::FunctionExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/GlobalExecutable.cpp:
(JSC::GlobalExecutable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/GlobalExecutable.h:
* Source/JavaScriptCore/runtime/InferredValue.h:
* Source/JavaScriptCore/runtime/InferredValueInlines.h:
(JSC::InferredValue<JSCellType>::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.cpp:
(JSC::JSFinalizationRegistry::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSFinalizationRegistry.h:
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSWeakObjectRef.cpp:
(JSC::JSWeakObjectRef::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/JSWeakObjectRef.h:
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::StructureTransitionTable::setSingleTransition):
(JSC::StructureTransitionTable::contains const):
(JSC::StructureTransitionTable::add):
(JSC::Structure::addNewPropertyTransition):
(JSC::Structure::removeNewPropertyTransition):
(JSC::Structure::attributeChangeTransition):
(JSC::Structure::nonPropertyTransitionSlow):
(JSC::Structure::visitChildrenImpl):
(JSC::Structure::setBrandTransition):
(JSC::Structure::finalizeUnconditionally):
(): Deleted.
(JSC::singleSlotTransitionWeakOwner): Deleted.
* Source/JavaScriptCore/runtime/Structure.h:
(JSC::Structure::shouldDoCacheableDictionaryTransitionForAdd):
(JSC::Structure::transitionCountEstimate const):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::addPropertyTransitionToExistingStructureImpl):
(JSC::StructureTransitionTable::trySingleTransition const):
(JSC::StructureTransitionTable::get const):
(JSC::StructureTransitionTable::finalizeUnconditionally):
(JSC::StructureTransitionTable::singleTransition const): Deleted.
* Source/JavaScriptCore/runtime/StructureRareData.cpp:
(JSC::StructureRareData::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/StructureRareData.h:
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
(JSC::StructureTransitionTable::~StructureTransitionTable):
(JSC::StructureTransitionTable::setMap):
(JSC::StructureTransitionTable::StructureTransitionTable): Deleted.
(JSC::StructureTransitionTable::weakImpl const): Deleted.
* Source/JavaScriptCore/runtime/SymbolTable.h:
* Source/JavaScriptCore/runtime/SymbolTableInlines.h:
(JSC::SymbolTable::finalizeUnconditionally):
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::WeakMapImpl<WeakMapBucket>::finalizeUnconditionally):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.cpp:
(JSC::JSWebAssemblyModule::finalizeUnconditionally):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.h:

Canonical link: https://commits.webkit.org/262809@main
